R/PipeOpFilterRows.R

```diff
@@ -0,0 +1,171 @@
+#'   that should be checked for missing values. If set to `all`, all columns of the data are used. A
+#'   row is removed if at least one missing value is found with respect to the columns specified.
+#'   Default is `NULL`, i.e., no removal of missing values is done.
+#' * `invert` :: `logical(1)`\cr
+#'   Should the filtering rule be set-theoretically inverted? Note that this happens after
+#'   (possible) missing values were removed if `na_column` is specified. Default is `FALSE`.
+#' * `skip_during_predict` :: `logical(1)`\cr
+#'   Should the filtering and missing value removal steps be skipped during prediction? If `TRUE`,
+#'   the input [`Task`][mlr3::Task] is returned unaltered during prediction. Default is `FALSE`.
+#'
+#' @section Internals:
+#' Uses the [`is.na()`][base::is.na] function for the checking of missing values.
+#'
+#' @section Methods:
+#' Only methods inherited from [`PipeOpTaskPreproc`]/[`PipeOp`].
+#'
+#' @examples
+#' library("mlr3")
+#' task = tsk("pima")
+#' po = PipeOpFilterRows$new(param_vals = list(
+#'   filter = expression(age < median(age) & mass > 30),
+#'   na_column = "all")
+#' )
+#' po$train(list(task))
+#' po$state
+#' @family PipeOps
+#' @include PipeOpTaskPreproc.R
+#' @export
+PipeOpFilterRows = R6Class("PipeOpFilterRows",
+  inherit = PipeOpTaskPreproc,
+  public = list(
+    initialize = function(id = "filterrows", param_vals = list()) {
+      ps = ParamSet$new(params = list(
+        ParamUty$new("filter", default = NULL, tags = c("train", "predict"), custom_check = function(x) {
+          ok = test_character(x, any.missing = FALSE, len = 1L) ||
+            is.expression(x) ||
+            test_integerish(x, lower = 1, min.len = 1L) ||
+            is.null(x)
+          if (!ok) return("Must either be a character vector of length 1, an expression, or an integerish object of row ids")
+          TRUE
+        }),
+        ParamUty$new("na_column", default = NULL, tags = c("train", "predict"), custom_check = function(x) {
+          check_character(x, any.missing = FALSE, min.len = 1L, null.ok = TRUE)
+        }),
+        ParamLgl$new("invert", default = FALSE, tags = c("train", "predict")),
+        ParamLgl$new("skip_during_predict", default = FALSE, tags = "predict"))
+      )
+      ps$values = list(filter = NULL, na_column = NULL, invert = FALSE, skip_during_predict = FALSE)
+      super$initialize(id, param_set = ps, param_vals = param_vals)
+    }
+  ),
+  private = list(
+    .na_and_filter = function(task, skip, set_state) {
+      if (skip) {
+        return(task)  # early exit if skipped (if skip_during_predict)
+      }
+
+      row_ids = task$row_ids
+
+      # NA column(s) handling
+      na = self$param_set$values$na_column
+      if (!is.null(na)) {
+        assert_subset(na, choices = c("all", colnames(task$data())))
+        if (na == "all") na = colnames(task$data())
+        na_ids = which(rowSums(is.na(task$data(cols = na))) > 0L)
```

Comment:
is this faster then `apply(..., ..., anyNA)` ?

Reply:
This should be faster as long as p is not >> n, which should be in most of the cases?
```
library(bench)

n <- 1000
p <- 1000
dat <- matrix(sample(c(NA, 0:p, NaN), n*p, TRUE), n, p)
# factor 3 for rowsums
mark(which(apply(dat, 1, anyNA)), which(rowSums(is.na(dat)) > 0L), iterations = 100)

n <- 10000
p <- 1000
dat <- matrix(sample(c(NA, 0:p, NaN), n*p, TRUE), n, p)
# factor 3.5 for rowsums
mark(which(apply(dat, 1, anyNA)), which(rowSums(is.na(dat)) > 0L), iterations = 100)

n <- 1000000
p <- 10
dat <- matrix(sample(c(NA, 0:p, NaN), n*p, TRUE), n, p)
# factor 23 for rowsums
mark(which(apply(dat, 1, anyNA)), which(rowSums(is.na(dat)) > 0L), iterations = 100)

n <- 10
p <- 100000
dat <- matrix(sample(c(NA, 0:p, NaN), n*p, TRUE), n, p)
# factor 1.8 for apply
mark(which(apply(dat, 1, anyNA)), which(rowSums(is.na(dat)) > 0L), iterations = 100)
```